clang/lib/CIR/CodeGen/CIRGenExprComplex.cpp

```diff
@@ -0,0 +1,947 @@
+    return Val.getComplexVal();
+
+  // If the lvalue is non-volatile, return the computed value of the assignment.
+  if (!LV.isVolatileQualified())
+    return Val.getComplexVal();
+
+  return buildLoadOfLValue(LV, E->getExprLoc());
+}
+
+mlir::Value ComplexExprEmitter::buildBinAdd(const BinOpInfo &Op) {
+  if (Op.LHS.getType().isa<mlir::cir::CIRFPTypeInterface>())
+    assert(!UnimplementedFeature::fpOptionsRaii());
+
+  return Builder.create<mlir::cir::BinOp>(
+      CGF.getLoc(Op.Loc), mlir::cir::BinOpKind::Add, Op.LHS, Op.RHS);
+}
+
+mlir::Value ComplexExprEmitter::buildBinSub(const BinOpInfo &Op) {
+  if (Op.LHS.getType().isa<mlir::cir::CIRFPTypeInterface>())
+    assert(!UnimplementedFeature::fpOptionsRaii());
+
+  return Builder.create<mlir::cir::BinOp>(
+      CGF.getLoc(Op.Loc), mlir::cir::BinOpKind::Sub, Op.LHS, Op.RHS);
+}
+
+mlir::Value ComplexExprEmitter::buildBinMul(const BinOpInfo &Op) {
+  if (Op.LHS.getType().isa<mlir::cir::CIRFPTypeInterface>())
+    assert(!UnimplementedFeature::fpOptionsRaii());
+
+  return Builder.create<mlir::cir::BinOp>(
+      CGF.getLoc(Op.Loc), mlir::cir::BinOpKind::Mul, Op.LHS, Op.RHS);
+}
+
+mlir::Value ComplexExprEmitter::buildBinDiv(const BinOpInfo &Op) {
+  if (Op.LHS.getType().isa<mlir::cir::CIRFPTypeInterface>())
+    assert(!UnimplementedFeature::fpOptionsRaii());
+
+  return Builder.create<mlir::cir::BinOp>(
+      CGF.getLoc(Op.Loc), mlir::cir::BinOpKind::Div, Op.LHS, Op.RHS);
+}
+
+LValue ComplexExprEmitter::buildBinAssignLValue(const BinaryOperator *E,
+                                                mlir::Value &Val) {
+  assert(CGF.getContext().hasSameUnqualifiedType(E->getLHS()->getType(),
+                                                 E->getRHS()->getType()) &&
+         "Invalid assignment");
+
+  // Emit the RHS.  __block variables need the RHS evaluated first.
+  Val = Visit(E->getRHS());
+
+  // Compute the address to store into.
+  LValue LHS = CGF.buildLValue(E->getLHS());
+
+  // Store the result value into the LHS lvalue.
+  buildStoreOfComplex(CGF.getLoc(E->getSourceRange()), Val, LHS,
+                      /*isInit*/ false);
+
+  return LHS;
+}
+
+mlir::Value ComplexExprEmitter::VisitBinAssign(const BinaryOperator *E) {
+  mlir::Value Val;
+  LValue LV = buildBinAssignLValue(E, Val);
+
+  // The result of an assignment in C is the assigned r-value.
+  if (!CGF.getLangOpts().CPlusPlus)
+    return Val;
+
+  // If the lvalue is non-volatile, return the computed value of the assignment.
+  if (!LV.isVolatileQualified())
+    return Val;
+
+  return buildLoadOfLValue(LV, E->getExprLoc());
+}
+
+mlir::Value ComplexExprEmitter::VisitBinComma(const BinaryOperator *E) {
+  CGF.buildIgnoredExpr(E->getLHS());
+  return Visit(E->getRHS());
+}
+
+mlir::Value ComplexExprEmitter::VisitChooseExpr(ChooseExpr *CE) {
+  return Visit(CE->getChosenSubExpr());
+}
+
+mlir::Value ComplexExprEmitter::VisitInitListExpr(InitListExpr *E) {
+  if (E->getNumInits() == 2) {
```

Comment:
Thanks for introducing the skeleton and for all the added testcases. However, this new file is quite big and introduces lots of CIRGen that I don't see tested (example, this one). Few options here:
- Add more testcases and exercise each of the visitors.
- Delete non-tested visitors and add `llvm_unreacheable` to them. Later on, do incremental patches to add more visitors and their correspondent tests.